pkg/util/ftp/ftp.go

```diff
@@ -0,0 +1,106 @@
+// Copyright 2024 Antrea Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ftp
+
+import (
+	"fmt"
+	"io"
+	"net/url"
+	"path"
+	"time"
+
+	"github.com/pkg/sftp"
+	"github.com/spf13/afero"
+	"golang.org/x/crypto/ssh"
+	"k8s.io/klog/v2"
+)
+
+const (
+	uploadToFileServerTries      = 5
+	uploadToFileServerRetryDelay = 5 * time.Second
+)
+
+func ParseFTPUploadUrl(uploadUrl string) (*url.URL, error) {
+	parsedURL, err := url.Parse(uploadUrl)
+	if err != nil {
+		return nil, err
+	}
+	if parsedURL.Scheme != "sftp" {
+		return nil, fmt.Errorf("not sftp protocol")
+	}
+	return parsedURL, nil
+}
+
+type Uploader interface {
+	// Upload uploads a file to the target sftp address using ssh config.
+	Upload(url string, fileName string, config *ssh.ClientConfig, outputFile afero.File) error
+}
+
+type SftpUploader struct {
+}
+
+func (uploader *SftpUploader) Upload(url string, fileName string, config *ssh.ClientConfig, outputFile afero.File) error {
+	if _, err := outputFile.Seek(0, 0); err != nil {
+		return fmt.Errorf("failed to upload to file server while setting offset: %v", err)
+	}
+	// url should be like: 10.92.23.154:22/path or sftp://10.92.23.154:22/path
+	parsedURL, _ := ParseFTPUploadUrl(url)
+	joinedPath := path.Join(parsedURL.Path, fileName)
+
+	triesLeft := uploadToFileServerTries
+	var uploadErr error
+	for triesLeft > 0 {
+		if uploadErr = upload(parsedURL.Host, joinedPath, config, outputFile); uploadErr == nil {
+			return nil
+		}
+		triesLeft--
+		if triesLeft == 0 {
+			return fmt.Errorf("failed to upload file after %d attempts", uploadToFileServerTries)
+		}
+		klog.InfoS("Failed to upload file", "UploadError", uploadErr, "TriesLeft", triesLeft)
+		time.Sleep(uploadToFileServerRetryDelay)
+	}
+	return nil
+}
+
+func upload(address string, path string, config *ssh.ClientConfig, file io.Reader) error {
```

Comment:
The patch ut coverage of this PR is 40%, better to add more unit test cases for these functions.

Reply:
i find it hard to mock ssl and sftp. Any suggestions?

Reply:
This should be covered by e2e testing BUT I have noticed that this PR did not update `pkg/controller/supportbundlecollection/controller.go` to use the `SFTPUploader`. Once this is addressed, maybe coverage for the patch will be better?

As for unit tests, I agree that it is not very practical to write one right now. As far as I can tell after taking a quick look, the `github.com/pkg/sftp` module does not make it easy for consumers to write unit tests. While we could create an SFTP server, it is not really suited for a unit test, as we cannot even provide a custom file system. We can leave it as future work.

Reply:
will update this.

Reply:
codecov/project increased after refactor support bundle controller to use SFTPUploader.